### ISSUE TITLE
observe in AppKit is crashing

### ISSUE THREAD

**Issue by zachwaugh** [pointfreeco/swift-composable-architecture#3297]
### Description

After the update to 1.13.1, `observe` is crashing in my macOS app. The actual error is:

```
Could not cast value of type 'SwiftNavigation.ObserveToken' (0x103ba9f88) to 'NSObject' (0x2010c88d0).
```

<img width="1426" alt="Screenshot 2024-08-22 at 3 43 28 PM" src="https://github.com/user-attachments/assets/57b4a304-aa08-4a29-a592-220395d7534e">

I'm assuming because `ObserveToken` in SwiftNavigation does not conform to `NSObject` - https://github.com/pointfreeco/swift-navigation/blob/main/Sources/SwiftNavigation/Observe.swift#L162. Where as the previous implementation of `observe` in TCA did use a class derived from `NSObject` https://github.com/pointfreeco/swift-composable-architecture/blob/1.12.1/Sources/ComposableArchitecture/UIKit/NSObject%2BObservation.swift#L200

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### The Composable Architecture version information

1.13.1

### Destination operating system

macOS 14.5

### Xcode version information

Xcode 15.4

**Comment by mbrandonw** [comment]
Hi @zachwaugh, I'm not able to reproduce this with the following test 

```swift
import AppKit
import XCTest

class MacTests: XCTestCase {
  @MainActor
  func testTokenStorage() {
    class Controller: NSViewController {
      override init(nibName nibNameOrNil: NSNib.Name?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        observe {}
      }

      required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
      }
    }

    let controller = Controller(nibName: nil, bundle: nil)
    _ = controller
  }
}
```

Can you provide a repro for us to look into?

**Comment by zachwaugh** [comment]
Interesting! I can't reproduce with that either, but noticed the one difference in my app is multiple `observe` calls and that seems to be the thing. This will reproduce it:

```swift
import AppKit
import XCTest

class MacTests: XCTestCase {
  @MainActor
  func testTokenStorage() {
    class Controller: NSViewController {
      override init(nibName nibNameOrNil: NSNib.Name?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        observe {}
        // 🚨 Crashes on second observe
        observe {}
      }

      required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
      }
    }

    let controller = Controller(nibName: nil, bundle: nil)
    _ = controller
  }
}
```

**Comment by mbrandonw** [comment]
Ah, interesting! That does reproduce for me. Thanks for that, will look into it more.

**Comment by mbrandonw** [comment]
This will be fixed once we merge and release https://github.com/pointfreeco/swift-navigation/pull/212/. If you want an immediate fix you could copy-paste the `observe` helper into your code and change the `Set<ObserveToken>` to an `[Any]`, then it will work.

**Comment by stephencelis** [comment]
This is fixed in `swift-navigation@2.1.0`.
